pkg/controller/plan/adapter/ova/builder.go

```diff
@@ -0,0 +1,594 @@
+					Source: &dvSource,
+					Storage: &cdi.StorageSpec{
+						Resources: core.ResourceRequirements{
+							Requests: core.ResourceList{
+								core.ResourceStorage: *resource.NewQuantity(diskSize, resource.BinarySI),
+							},
+						},
+						StorageClassName: &storageClass,
+					},
+				}
+				// set the access mode and volume mode if they were specified in the storage map.
+				// otherwise, let the storage profile decide the default values.
+				if mapped.Destination.AccessMode != "" {
+					dvSpec.Storage.AccessModes = []core.PersistentVolumeAccessMode{mapped.Destination.AccessMode}
+				}
+				if mapped.Destination.VolumeMode != "" {
+					dvSpec.Storage.VolumeMode = &mapped.Destination.VolumeMode
+				}
+
+				dv := dvTemplate.DeepCopy()
+				dv.Spec = dvSpec
+				if dv.ObjectMeta.Annotations == nil {
+					dv.ObjectMeta.Annotations = make(map[string]string)
+				}
+				dv.ObjectMeta.Annotations[planbase.AnnDiskSource] = getDiskSourcePath(disk.FilePath, disk.Name)
+				dvs = append(dvs, *dv)
+			}
+		}
+	}
+
+	return
+}
+
+// Create the destination Kubevirt VM.
+func (r *Builder) VirtualMachine(vmRef ref.Ref, object *cnv.VirtualMachineSpec, persistentVolumeClaims []core.PersistentVolumeClaim) (err error) {
+	vm := &model.VM{}
+	err = r.Source.Inventory.Find(vm, vmRef)
+	if err != nil {
+		err = liberr.Wrap(
+			err,
+			"VM lookup failed.",
+			"vm",
+			vmRef.String())
+		return
+	}
+
+	if r.Plan.Spec.Warm {
+		err = liberr.New(
+			fmt.Sprintf(
+				"Warm migration is disabled for VM %s from OVA provider",
+				vmRef.String()))
+		return
+	}
+
+	var conflicts []string
+	conflicts, err = r.macConflicts(vm)
+	if err != nil {
+		return
+	}
+	if len(conflicts) > 0 {
+		err = liberr.New(
+			fmt.Sprintf("Source VM has a mac address conflict with one or more destination VMs: %s", conflicts))
+		return
+	}
+
+	if object.Template == nil {
+		object.Template = &cnv.VirtualMachineInstanceTemplateSpec{}
+	}
+	r.mapDisks(vm, persistentVolumeClaims, object)
+	r.mapFirmware(vm, object)
+	r.mapCPU(vm, object)
+	r.mapMemory(vm, object)
+	r.mapClock(object)
+	r.mapInput(object)
+	err = r.mapNetworks(vm, object)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func (r *Builder) mapNetworks(vm *model.VM, object *cnv.VirtualMachineSpec) (err error) {
+	var kNetworks []cnv.Network
+	var kInterfaces []cnv.Interface
+
+	numNetworks := 0
+	netMapIn := r.Context.Map.Network.Spec.Map
+	for i := range netMapIn {
+		mapped := &netMapIn[i]
+		ref := mapped.Source
+		network := &model.Network{}
+		fErr := r.Source.Inventory.Find(network, ref)
+		if fErr != nil {
+			err = fErr
+			return
+		}
+
+		needed := []ova.NIC{}
+		for _, nic := range vm.NICs {
+			if nic.Network == network.Name {
+				needed = append(needed, nic)
+			}
+		}
+		if len(needed) == 0 {
+			continue
+		}
+		for _, nic := range needed {
+			networkName := fmt.Sprintf("net-%v", numNetworks)
+			numNetworks++
+			if len(nic.MAC) == 0 {
+				nic.MAC = generateRandomMACAddress()
```

Comment:
do we have to generate one? (doesn't CNV generate MAC address when it's not specified?)